configuration_go/k8sutil/v2/providers.go

```diff
@@ -0,0 +1,430 @@
+	}
+}
+
+func (c *ConfigMap) MakeManifest() runtime.Object {
+	return &corev1.ConfigMap{
+		TypeMeta:   k8sutil.ConfigMapMeta,
+		ObjectMeta: c.MetaConfig.MakeMeta(),
+		Data:       c.Data,
+	}
+}
+
+type Secret struct {
+	MetaConfig
+	Data map[string][]byte
+}
+
+func NewSecret(metaConfig MetaConfig, data map[string][]byte) *Secret {
+	return &Secret{
+		MetaConfig: metaConfig,
+		Data:       data,
+	}
+}
+
+func (s *Secret) MakeManifest() runtime.Object {
+	return &corev1.Secret{
+		TypeMeta:   k8sutil.SecretMeta,
+		ObjectMeta: s.MetaConfig.MakeMeta(),
+		Data:       s.Data,
+	}
+}
+
+// ContainerProvider is the interface that containers must implement to be added to a pod.
+type ContainerProvider interface {
+	GetContainer() corev1.Container
+	GetVolumes() []corev1.Volume
+
+	ServiceProvider
+	ServiceMonitorProvider
+	PersistentVolumeClaimProvider
+	ManifestsProvider
+}
+
+// Container represents a container in a pod.
+// It implements the ContainerProvider interface.
+type Container struct {
+	Name            string
+	Image           string
+	ImageTag        string
+	ImagePullPolicy corev1.PullPolicy
+	Resources       corev1.ResourceRequirements
+	Env             []corev1.EnvVar
+	LivenessProbe   *corev1.Probe
+	ReadinessProbe  *corev1.Probe
+	Args            []string
+	Ports           []corev1.ContainerPort
+	VolumeMounts    []corev1.VolumeMount
+
+	Volumes      []corev1.Volume
+	VolumeClaims []VolumeClaim
+	ServicePorts []corev1.ServicePort
+	MonitorPorts []monv1.Endpoint
+	Manifests    map[string]ManifestProvider
+}
+
+// GetContainer returns a Kubernetes Container.
+func (c *Container) GetContainer() corev1.Container {
+	return corev1.Container{
+		Name:                     c.Name,
+		Image:                    fmt.Sprintf("%s:%s", c.Image, c.ImageTag),
+		ImagePullPolicy:          c.ImagePullPolicy,
+		Resources:                c.Resources,
+		Env:                      c.Env,
+		LivenessProbe:            c.LivenessProbe,
+		ReadinessProbe:           c.ReadinessProbe,
+		Args:                     c.Args,
+		Ports:                    c.Ports,
+		TerminationMessagePolicy: corev1.TerminationMessageFallbackToLogsOnError,
+		VolumeMounts:             c.VolumeMounts,
+	}
+}
+
+// GetVolumes returns the volumes that the container requires.
+func (c *Container) GetVolumes() []corev1.Volume {
+	return c.Volumes
+}
+
+// GetServicePorts returns the ports that the container exposes.
+func (c *Container) GetServicePorts() []corev1.ServicePort {
+	return c.ServicePorts
+}
+
+// GetServiceMonitorEndpoints returns the endpoints to be monitored by Prometheus.
+func (c *Container) GetServiceMonitorEndpoints() []monv1.Endpoint {
+	return c.MonitorPorts
+}
+
+// GetVolumeClaims returns the volume claims that the container requires.
+func (c *Container) GetVolumeClaims() []VolumeClaim {
+	return c.VolumeClaims
+}
+
+// AddConfigMap adds a config map to the container.
+func (c *Container) AddManifest(filename string, mf ManifestProvider) {
```

Comment:
Missmatching name according to signature and function's docs. `AddConfigMap` seems to be correct.

Reply:
Yup there is a mismatch. The objective here is to make the container self contained with everything it needs to work properly, including configmaps and secrets. To do that, it implements the `ManifestsProvider` interface. Here this is a function to add any type of manifest with its expected file name. `AddManifest` is what I wanted. 
Updating.

Reply:
I changed it to handle config maps.